### ISSUE TITLE
Port LogContext to .NET Core

### ISSUE THREAD

**Issue by nblumhardt** [serilog/serilog#588]
From #535. This may use AsyncLocal<T>.

**Comment by nblumhardt** [comment]
Ready to attempt on `dev` if anyone is interested.

**Comment by colin-young** [comment]
What's the status on this? Still up for grabs? I might be able to make an attempt at it, especially seeing as I'd kind of like to make use of it.

**Comment by nblumhardt** [comment]
Absolutely! I'll knock off the `up-for-grabs` label in anticipation :-) ... Let us know if you need a hand with any of it.

**Comment by colin-young** [comment]
Fork and pull request the usual way to do things here? I'm sure I'll be back with questions.

**Comment by colin-young** [comment]
I'm presuming from the initial comment here that the intention was to replace CallContext with AsyncLocal<T> for .Net 4.5 and up?

**Comment by nblumhardt** [comment]
I think so - `AsyncLocal` looks like the right API going forwards :+1:

**Comment by colin-young** [comment]
So I've got things building now, with `net452`, `dotnet5.1` and `dotnet5.4`. My biggest concern, which I'm just about to start investigating is the unit tests. VS Test Explorer gives me 160 tests passed, 1 skipped. Build.ps1 gives me a total of 76 tests, 5 failing, 0 skipped, but the the console output clearly shows it is skipping `DoesNotPrevenCrossDomainCalls` because it "fails when run from Build.ps1."

Those numbers clearly don't agree with each other, so I have very little confidence in the current state of my fork.

Figured I'd ask here in case this is something that's sort of expected. In the meantime, I'm going to see if I can make sense of things.

**Comment by khellang** [comment]
@colin-young It depends on the targets that you're running the tests under. Some of them are #ifdef'ed for `NET45` etc., so the numbers don't necessarily add up :smile:

**Comment by colin-young** [comment]
@khellang Yup. Eventually figured that out when I decided to scroll back far enough in the console output to see the DNX 4.5.2 tests. I think you mean #ifdef'd as features, not platform :)

I was more concerned about the failure from the command line and having it passing in VS. The failure was a LogContext implementation error by me (surprise!) that I've corrected and now all tests are passing... however, I'm still concerned about the fact that VS somehow made that one pass with the error.

I'm going to do a bit of testing of the new-to-core LogContext in my test project, and then see if I can't figure out some way to do some async unit testing to make sure it really works.
